os/drivers/serial/serial.c

```diff
@@ -125,6 +139,86 @@ static const struct file_operations g_serialops = {
  * Private Functions
  ************************************************************************************/
 
+#ifdef CONFIG_DYNAMIC_LOGS_STORING
+void read_log_net(void)
+{	size_t free_size;
+
+	if (!sq_empty(&log_net_chunks)) { /* check if there are any logs saved */
+		struct mm_heap_s *myheap = mm_get_heap(sq_peek(&log_net_chunks)); /* get heap details using the head of the chunks linked list */
+		if (myheap == NULL) {
+			lldbg("mm_get_heap failed\n");
+			return;
+		}
+		free_size = myheap->mm_heapsize - myheap->total_alloc_size; /* get the free size */
+	} else { /* we have not logs stored, so return */
+		return;
+	}
+
+	lldbg("Read output ---> starts here <--- log size : %d, free heap : %d\n", log_net_size, free_size);
+
+	while (!sq_empty(&log_net_chunks)) {
+		struct log_net_chunk_s *next_chunk = (struct log_net_chunk_s *)sq_remfirst(&log_net_chunks);
+		for (int i = 0; i < next_chunk->curbytes; i++) { /* printing the data to verify storing of logs, later to be replaced by network code to send to server*/
+                        lldbg_noarg("%c", next_chunk->arr[i]);
+                }
+
+                kmm_free(next_chunk); /* once we read the stored data, free the node */
+        }
+
+	lldbg("Read output ---> ends here <---\n");
+}
+
+static void add_char_to_log_net(char ch)
+{
+	if (!log_net_chunks_initialized) { /* one time queue init */
+		sq_init(&log_net_chunks);
+		log_net_chunks_initialized = true;
+	}
+
+	if (sq_empty(&log_net_chunks)) {
+		/* no chunks are present so we allocate new chunk */
+		struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
+		node->curbytes = 1;
+		node->arr[0] = ch;
+		sq_addlast((sq_entry_t *)node, &log_net_chunks);
+		/* set the size of chunks to 1 chunk */
+		log_net_size = sizeof(struct log_net_chunk_s);
+	} else {
+		/* need to check if the current chunks size if over max_log_size or greater than x% of free heap */
+		struct mm_heap_s *myheap = mm_get_heap(sq_peek(&log_net_chunks)); /* get heap details using the head of the chunks linked list */
+		if (!myheap) {
+			lldbg("mm_get_heap failed\n");
+			return; /* Do nothing */
+		}
+		size_t free_size = myheap->mm_heapsize - myheap->total_alloc_size; /* get the free size */
+
+		while (!sq_empty(&log_net_chunks) && (log_net_size / free_size * 100 > CONFIG_MAX_FREE_HEAP_PERCENT || log_net_size > CONFIG_MAX_LOG_SIZE)) {
+			struct log_net_chunk_s *node = (struct log_net_chunk_s *)sq_remfirst(&log_net_chunks);
+			kmm_free(node);
+			lldbg("\n node removed \n");
+			log_net_size -= sizeof(struct log_net_chunk_s);
+		}
```

Comment:
How about changing to calculate how many chunk should be released instead of checking one by one?